diff --git a/xbmc-pvr-addons/configure.ac b/xbmc-pvr-addons/configure.ac
--- a/xbmc-pvr-addons/configure.ac
+++ b/xbmc-pvr-addons/configure.ac
@@ -25,9 +25,9 @@
 ### External dependencies option
 AC_ARG_ENABLE([addons-with-dependencies],
   [AS_HELP_STRING([--enable-addons-with-dependencies],
-  [enable build of addons that depend on external libraries (default is no)])],
+  [enable build of addons that depend on external libraries (default is yes)])],
   [build_addons_with_dependencies=$enableval],
-  [build_addons_with_dependencies=no])
+  [build_addons_with_dependencies=yes])
 
 AC_ARG_ENABLE([mysql],
   [AS_HELP_STRING([--disable-mysql],
   
diff --git a/xbmc-pvr-addons/configure.ac b/xbmc-pvr-addons/configure.ac
index 0f8766d..e99b373 100644
--- a/xbmc-pvr-addons/configure.ac
+++ b/xbmc-pvr-addons/configure.ac
@@ -211,7 +211,7 @@ fi
 ### Addons with dependencies
 AM_CONDITIONAL([ADDON_MYTHTV], [test "$build_addons_with_dependencies" = "yes" -a "$use_mysql" = "yes" -a "$use_boost" = "yes"])
 AM_CONDITIONAL([ADDON_IPTVSIMPLE], [test "$build_addons_with_dependencies" = "yes" -a "$use_zlib" = "yes"])
-AM_CONDITIONAL([ADDON_FILMON], [test "$build_addons_with_dependencies" = "yes" -a "$use_curl" = "yes"])
+AM_CONDITIONAL([ADDON_FILMON], [test "$build_addons_with_dependencies" = "yes" -a "$use_curl" = "yes" -a "$use_jsoncpp" = "yes" -a "$use_cryptopp" = "yes"])
 
 ### Check for Intree building
 if test "x${cross_compiling}" = "xyes" || test "x${cross_compiling}" = "xmaybe"; then
