diff --git a/xbmc-pvr-addons/configure.ac b/xbmc-pvr-addons/configure.ac
--- a/xbmc-pvr-addons/configure.ac
+++ b/xbmc-pvr-addons/configure.ac
@@ -25,9 +25,9 @@
 ### External dependencies option
 AC_ARG_ENABLE([addons-with-dependencies],
   [AS_HELP_STRING([--enable-addons-with-dependencies],
-  [enable build of addons that depend on external libraries (default is no)])],
+  [enable build of addons that depend on external libraries (default is yes)])],
   [build_addons_with_dependencies=$enableval],
-  [build_addons_with_dependencies=no])
+  [build_addons_with_dependencies=yes])
 
 AC_ARG_ENABLE([mysql],
   [AS_HELP_STRING([--disable-mysql],
   
diff --git a/xbmc-pvr-addons/configure.ac b/xbmc-pvr-addons/configure.ac
--- a/xbmc-pvr-addons/configure.ac
+++ b/xbmc-pvr-addons/configure.ac
@@ -293,18 +293,18 @@
 if test "x${libdir}" != 'x${exec_prefix}/lib'; then
   LIBDIR=${libdir}
 elif test "$prefix" = "NONE"; then
-  LIBDIR=$PREFIX_DEFAULT/lib/xbmc/addons
+  LIBDIR=$PREFIX_DEFAULT/lib/kodi/addons
 else
-  LIBDIR=$prefix/lib/xbmc/addons
+  LIBDIR=$prefix/lib/kodi/addons
 fi
 AC_SUBST(LIBDIR)
 
 if test "x${datadir}" != 'x${datarootdir}' && test "x${datarootdir}" = 'x${prefix}/share'; then
   DATADIR=${datadir}
 elif test "$prefix" = "NONE"; then
-  DATADIR=$PREFIX_DEFAULT/share/xbmc/addons
+  DATADIR=$PREFIX_DEFAULT/share/kodi/addons
 else
-  DATADIR=$prefix/share/xbmc/addons 
+  DATADIR=$prefix/share/kodi/addons 
 fi
 AC_SUBST(DATADIR)